Automatically generate a single yaml file that users can download to simplify the operator deployment

whowutwut opened the pull request:
This PR helps to fix issues brought up in #74, and does the following: 

- [x]  Concat the following yaml files to a single one  called: `ibm-spectrum-scale-csi-operator.yaml`
   ```
    operator_files: 
      - name: "deploy/operator.yaml"
      - name: "deploy/role.yaml"
      - name: "deploy/role_binding.yaml"
      - name: "deploy/service_account.yaml"
      - name: "deploy/crds/ibm_v1alpha1_csiscaleoperator_crd.yaml"
   ```
- [x] Test driven, show the failure.  Add a travisCI test that is invoked by: `ansible-playbook  ansible/generate-playbook.yaml --extra-vars "travis_testing=True"` that will create a test file and diff it.  If it doesn't match what's checked in, it means that someone in the deploy/*.yaml files have  changed and was not updated in the PR.  This is to help us catch the problems

- [x] Modify the documentation 

  After some searching, I decided that RST does provide us more rendered look and features complete when compared to simple markdown.  I think the direction is really Asciidoctor, but this is not in line with RTD roadmap, so we would have to host/build it ourselves. Since that's not viable, I rewrote in RST and tried to set up some good practices structure in the `docs/source` area. 

---------

The development team would do the following to create and update the files.  Then if any changes, it  should be commit from their  forked copy.  The PR should catch the  missing commit of the file. 

```
cd ibm-spectrum-scale-csi-operator

ansible-playbook  ansible/generate-playbook.yaml 
```

The generated file, would be created
![image](https://user-images.githubusercontent.com/10049070/69578581-cb671000-0f9e-11ea-8f35-87ad83b8e68a.png)


## Testing 

Without the file checked in... 
<img width="1331" alt="image" src="https://user-images.githubusercontent.com/10049070/69578871-55af7400-0f9f-11ea-9021-4874a1d6a84a.png">

## New Flow 

1. User curls 2 files
   * namespace.yaml
   * ibm-spectrum-scale-csi-operator.yaml

2. `kubctl/oc apply -f ` of those 2 files, which brings up the operator pods  

3.  Curl down the `example/spectrum_scale.yaml` and modify to their cluster and apply it to bring up the driver pods.

whowutwut commented:
With the following commit eda621f717003dafcf050310d5f5206b3f4db867, printing out standard error:

```

TASK [[Testing] Diff standard error] **************************************************************************************************************************************
ok: [localhost] => {
    "msg": "diff: /Users/vhu/Documents/Development/GitHub/forked/ibm-spectrum-scale-csi-operator/ansible/..//generated/installer/ibm-spectrum-scale-csi-operator.yaml: No such file or directory"
}
```

whowutwut commented:
Here's what the success case for TravisCI looks like 

<img width="1085" alt="image" src="https://user-images.githubusercontent.com/10049070/69579466-880da100-0fa0-11ea-9f98-70d8cf187a38.png">

whowutwut commented on ansible/generate-playbook.yaml:
```diff
@@ -0,0 +1,27 @@
+    repo_root_dir: "{{ playbook_dir }}/../"
+    operator_dir: "{{ repo_root_dir }}/stable/ibm-spectrum-scale-csi-operator-bundle/operators/ibm-spectrum-scale-csi-operator"
+
+    generated_dir: "{{ repo_root_dir }}/generated/installer"
```
So here's a potential question, I'm generating this file at the repo root directory, but in reality all the deployment scripts are under `/stable/ibm-spectrum-scale-csi-operator-bundle/operators/ibm-spectrum-scale-csi-operator`, so should the generated files be there? 

That means the curl command would be crazy long, something like: 

https://raw.githubusercontent.com/IBM/ibm-spectrum-scale-csi-operator/master/stable/ibm-spectrum-scale-csi-operator-bundle/operators/ibm-spectrum-scale-csi-operator/generated/installer/ibm-spectrum-scale-csi-operator.yaml

I guess it doesn't matter that length...  thoughts?

mew2057 replied:
I think this will depend on whether we keep the CASE format in the git repo,  since this is  an arbitrary  path for CloudPak  cert, I think an argument can be made that we might want to flatten this out and have  something  that collates the operator and driver in the near future. I think  there's a need to discuss this however.

mew2057 replied:
I think that we should set this as a 1.0.1 target, because I foresee a change in how we handle cloudpak directory organization.

whowutwut commented:
This PR is complete and ready for review.

whowutwut commented on docs/source/deployment/manual.rst:
```diff
@@ -0,0 +1,71 @@
+
+    curl https://raw.githubusercontent.com/IBM/ibm-spectrum-scale-csi-operator/master/generated/installer/namespace.yaml > namespace.yaml
+    curl https://raw.githubusercontent.com/IBM/ibm-spectrum-scale-csi-operator/master/generated/installer/ibm-spectrum-scale-csi-operator.yaml > ibm-spectrum-scale-csi-operator.yaml
+    curl https://raw.githubusercontent.com/IBM/ibm-spectrum-scale-csi-operator/master/generated/installer/examples/spectrum_scale.yaml > spectrum_scale.yaml
```
This points at `master` which is a little problematic if we want users  to look at docs at a specific version and be able to `copy/paste` these samples.  

This may  be OK, if we ensure master is always stable/tested... and a release quality, which today  it is not.

mew2057 replied:
yeah, we're probably going to want to change this to be something release specific.

whowutwut commented:
Don't merge this yet, I am trying to do some testing and `apply -f ` of this yaml files doesn't bring anything up.   So will continue to test this next week.

whowutwut commented:
Ok turns out the issue is that we need to have `---` separating the blocks of yaml files, I  modified the generation code to create separators for the blocks, after that, I am able to bring up the operator with the namespace + single file 

```
[root@c943f5n01-pvt csi-operator-ansible]# oc apply -f curl.tmp/namespace.yaml
namespace/ibm-spectrum-scale-csi-driver created


[root@c943f5n01-pvt csi-operator-ansible]# oc apply -f curl.tmp/ibm-spectrum-scale-csi-operator.yaml
deployment.apps/ibm-spectrum-scale-csi-operator created
role.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-operator created
clusterrole.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-operator created
clusterrole.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-node created
clusterrole.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-attacher created
clusterrole.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-provisioner created
rolebinding.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-operator created
clusterrolebinding.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-operator created
clusterrolebinding.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-node created
clusterrolebinding.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-provisioner created
clusterrolebinding.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-attacher created
serviceaccount/ibm-spectrum-scale-csi-operator created
serviceaccount/ibm-spectrum-scale-csi-attacher created
serviceaccount/ibm-spectrum-scale-csi-node created
serviceaccount/ibm-spectrum-scale-csi-provisioner created
customresourcedefinition.apiextensions.k8s.io/csiscaleoperators.scale.ibm.com created


[root@c943f5n01-pvt csi-operator-ansible]# ./operator-helper.sh get_pods
+ oc get pods -n ibm-spectrum-scale-csi-driver
NAME                                               READY   STATUS              RESTARTS   AGE
ibm-spectrum-scale-csi-operator-6ff9cf6979-hk2hj   0/2     ContainerCreating   0          2s
+ set +x
[root@c943f5n01-pvt csi-operator-ansible]# ./operator-helper.sh get_pods
+ oc get pods -n ibm-spectrum-scale-csi-driver
NAME                                               READY   STATUS              RESTARTS   AGE
ibm-spectrum-scale-csi-operator-6ff9cf6979-hk2hj   0/2     ContainerCreating   0          5s
+ set +x
[root@c943f5n01-pvt csi-operator-ansible]# ./operator-helper.sh get_pods
+ oc get pods -n ibm-spectrum-scale-csi-driver
NAME                                               READY   STATUS              RESTARTS   AGE
ibm-spectrum-scale-csi-operator-6ff9cf6979-hk2hj   0/2     ContainerCreating   0          8s
+ set +x
[root@c943f5n01-pvt csi-operator-ansible]# ./operator-helper.sh get_pods
+ oc get pods -n ibm-spectrum-scale-csi-driver
NAME                                               READY   STATUS    RESTARTS   AGE
ibm-spectrum-scale-csi-operator-6ff9cf6979-hk2hj   2/2     Running   0          13s
+ set +x
[root@c943f5n01-pvt csi-operator-ansible]#
```

whowutwut commented:
Awesome, the TravisCI was failing from this: 

<img width="1085" alt="image" src="https://user-images.githubusercontent.com/10049070/69997300-d2070180-1521-11ea-9353-eca873e833d9.png">

This is caused by the name change of the CRD yami file ....  I rebased this branch to update the latest  in  `dev` .., when I ran the ansible playbook to test, it fails 

```

TASK [[Testing] Diff standard error] *****************************************************************************************************************
ok: [localhost] => {
    "msg": ""
}

TASK [[Testing] Failure Message] *********************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "The generated ibm-spectrum-scale-csi-operator.yaml did not match the deploy yaml files, did you forget to check it in?"}

PLAY RECAP *******************************************************************************************************************************************
localhost                  : ok=9    changed=4    unreachable=0    failed=1    skipped=2    rescued=0    ignored=1   

victors-mbp-2:ibm-spectrum-scale-csi-operator vhu$
```

Which is expected...  made fixes for it with dcfd1b51673e99ca08a9c4d229725c0c67e06173

whowutwut commented:
@yadaven @mew2057 @aspalazz @gandhisanjayv Few things we need to sync up on and I can make sure this PR is ready...  

1.  Is `ibm-spectrum-scale-csi-operator.yaml` an OK name for the combined yaml file? 
2. Is https://github.com/IBM/ibm-spectrum-scale-csi-operator/generated/installer a good location to put these files? 
3. Should we also copy the namespace file over to the `generated/installer` path? (maybe not...)

mew2057 commented:
1.  That should be fine.
2. Also should be fine.
3. Probably not, I don't really see a point.

whowutwut commented:
If everyone is OK with Johns response, then that reflects what's in this PR, I don't plan to make any additional changes, I would like to request that this PR gets into GA.  The merge will render the documentation which the URLs are pointing at master...  maybe OK for now..

gandhisanjayv commented:
how do we plan to package csi-driver files?
generate_pv_yaml.sh,spectrum-scale-driver-snap.sh and examples

whowutwut commented:
i don't think we need to package those files as part of the operator.  I do think we should remove the examples and have a single CR, like Matt mentioned on the call..

whowutwut commented:
This is kind of scoping out, but I think I need to figure the single CR file before we merge this one...  instead of pulling examples, and document it as such...    right now the docs point as downloading an `example/spectrum_scale.yaml` which deploys the driver.

whowutwut commented:
Rebased this branch to pull in the latest changes in the `dev` branch, and did the following: 
* minor changes to  the formatting of the comments in the single CR file
* updated to remove the paths pointing to the example directory, which is removed
* changes for the RTD documentation

whowutwut commented:
Travis testing now successful with that last commit. Individual yaml files have changed which caused travis to fail and flag that missing update, working as designed. 👍
